states/oklahoma/parse_pdf.py

```diff
@@ -0,0 +1,110 @@
+from common import cache_request
+from io import BytesIO
+from pdfminer.pdfinterp import PDFResourceManager, PDFPageInterpreter
+from pdfminer.converter import TextConverter
+from pdfminer.layout import LAParams
+from pdfminer.pdfpage import PDFPage
+from io import StringIO
+import os
+from collections import defaultdict
+
+
+def get_text(url):
+  req = cache_request(url, is_binary=True)
+  with BytesIO(req) as fh:
+    output = StringIO()
+    manager = PDFResourceManager()
+    converter = TextConverter(manager, output, laparams=LAParams())
+    interpreter = PDFPageInterpreter(manager, converter)
+
+    for page in PDFPage.get_pages(fh, set()):
+      interpreter.process_page(page)
+    
+    converter.close()
+    text = output.getvalue()
+    output.close()
+    return text
+
+def makes_lines(text):
+  lines = []
+  line = ''
+  for character in text:
+    if character == '\n':
+      lines.append(line)
+      line = ''
+      continue
+    line += character
+
+  # remove header
+  return lines[5:]
+
+
+def group_rows(lines):
+  add_data = False
```

Comment:
If you have the lines, the easier solution is to just look for the headers (e.g. `COUNTY`), then consume lines until you get to an empty or frivolous one and set that as a column.  You can then align the columns using `zip` (you have to assume they're alined here).

Sorry, I thought you were reading it row by row and had to pull this out to see what you were doing.

The problem with logic like this is that it's hard to tell by reading if the magic numbers are correct.

Reply:
I see what you're saying.  I agree about the magic numbers, but I think that the way I have it now is better than using zip (fewer list variables).  Let me know if you still disagree and I'd be happy to change it.